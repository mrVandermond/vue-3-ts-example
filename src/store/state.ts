import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';

export interface IState {
  favoritePhotos: IPhotoWithIsFavorite[];
  modal: IModalConfig;
  title: ITitleConfig;
}

export const state = (): IState => ({
  favoritePhotos: [],
  modal: {
    isOpen: false,
    url: '',
  },
  title: {
    isShow: false,
    text: '',
    xCoord: 0,
    yCoord: 0,
  },
});
