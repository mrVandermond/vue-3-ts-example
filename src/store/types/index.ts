/* mutations */
export enum EMutations {
  ADD_TO_FAVORITES = 'ADD_TO_FAVORITES',
  REMOVE_FROM_FAVORITES = 'REMOVE_FROM_FAVORITES',
  SET_MODAL_CONFIG = 'SET_MODAL_CONFIG',
  SET_TITLE_CONFIG = 'SET_TITLE_CONFIG',
}
/* mutations */
/* getters */
export enum EGetters {
  IS_FAVORITE_PHOTO_BY_ID = 'IS_FAVORITE_PHOTO_BY_ID',
}
/* getters */
/* actions */
export enum EActions {
  GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE = 'GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE',
  SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE = 'SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE',
  FETCH_USERS = 'FETCH_USERS',
  FETCH_ALBUMS_BY_USER_ID = 'FETCH_ALBUMS_BY_USER_ID',
  FETCH_PHOTOS_BY_ALBUM_ID = 'FETCH_PHOTOS_BY_ALBUM_ID',
}
/* actions */

export const storeTypes = {
  EMutations,
  EGetters,
  EActions,
};
