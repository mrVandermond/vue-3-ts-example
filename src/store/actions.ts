import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
