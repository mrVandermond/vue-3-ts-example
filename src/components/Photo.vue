<template>
  <div
    class="photo"
    :data-title="props.photo.title"
    @click="onClickOpenModal"
  >
    <PhotoStar
      :is-active="props.photo.isFavorite"
      class="photo-star"
      @click.stop="onClickStar"
    />

    <img
      :src="props.photo.thumbnailUrl"
      alt="photo-image"
    >
  </div>
</template>

<script lang="ts" setup>
import PhotoStar from '@/components/PhotoStar.vue';
import type { IPhotoWithIsFavorite } from '@/types';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const props = defineProps<{
  photo: IPhotoWithIsFavorite;
}>();
const emit = defineEmits<{
  (event: 'addedToFavorites', photo: IPhotoWithIsFavorite): void;
  (event: 'removeFromFavorites', photo: IPhotoWithIsFavorite): void;
}>();
const store = customUseStore();

const onClickStar = (): void => {
  if (props.photo.isFavorite) {
    emit('removeFromFavorites', props.photo);

    return;
  }

  emit('addedToFavorites', props.photo);
};
const onClickOpenModal = (): void => {
  store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
    isOpen: true,
    url: props.photo.url,
  });
};
</script>

<style lang="scss">
.photo {
  overflow: hidden;
  border-radius: 4px;
  width: 150px;
  height: 150px;
  cursor: pointer;
  position: relative;
}
.photo-star {
  position: absolute;
  top: 8px;
  right: 8px;
}
</style>
