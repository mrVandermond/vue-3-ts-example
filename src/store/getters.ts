import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
