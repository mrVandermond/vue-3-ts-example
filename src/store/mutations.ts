import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
