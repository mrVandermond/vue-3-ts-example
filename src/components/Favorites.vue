<template>
  <div
    :class="['inner-container', {
      'inner-container__align-start': !isEmptyFavorites
    }]"
  >
    <div
      v-if="isEmptyFavorites"
      class="empty-block"
    >
      <img
        src="assets/empty.png"
        class="empty-image"
        alt="empty-image"
      >

      <p class="title-text title-text__bold">
        Список избранного пуст
      </p>
      <p class="title-text">
        Добавляйте изображения, нажимая на звездочки
      </p>
    </div>

    <div
      v-else
      class="list-photo"
    >
      <Photo
        v-for="photo in favoritePhotos"
        :key="photo.id"
        :photo="photo"
        @remove-from-favorites="onRemoveFromFavorites"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed, onBeforeMount, onBeforeUnmount } from 'vue';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';
import type { IPhotoWithIsFavorite } from '@/types';
import Photo from '@/components/Photo.vue';

const store = customUseStore();
const favoritePhotos = computed(() => store.state.favoritePhotos);
const isEmptyFavorites = computed(() => !favoritePhotos.value.length);

onBeforeMount(async (): Promise<void> => {
  if (store.state.favoritePhotos.length) return;

  await store.dispatch(storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE);
});
onBeforeUnmount(async (): Promise<void> => {
  await store.dispatch(storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE);
});

const onRemoveFromFavorites = (photo: IPhotoWithIsFavorite): void => {
  store.commit(storeTypes.EMutations.REMOVE_FROM_FAVORITES, photo);
};
</script>

<style lang="scss">
.inner-container {
  margin-top: 56px;
  min-height: calc(100vh - 64px - 53px - 56px);

  &__align-start {
    align-content: flex-start;
  }
}
.empty-block {
  text-align: center;
}
.empty-image {
  margin-bottom: 32px;
}
.list-photo {
  display: flex;
  flex-wrap: wrap;
  gap: 42px;
  width: 534px;
}
</style>
