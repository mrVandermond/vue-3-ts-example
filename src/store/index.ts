import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
