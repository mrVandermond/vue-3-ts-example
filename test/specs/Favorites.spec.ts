import Favorites from '@/components/Favorites.vue';
import Photo from '@/components/Photo.vue';

import type { Store } from '@/store/types/overrides';
import type { IPhotoWithIsFavorite } from '@/types';

import { key, customCreateStore } from '@/store';
import { storeTypes } from '@/store/types';
import { flushPromises } from '@vue/test-utils';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';

import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

const EXAMPLE_PHOTO: IPhotoWithIsFavorite = {
  isFavorite: true,
  albumId: 1,
  url: 'test_url',
  id: 1,
  title: 'test_title',
  thumbnailUrl: 'test_thumbnail_url',
};

describe('Favorites component', () => {
  let store!: Store;
  const getWrapper = (): TWrapperFactoryReturnType => wrapperFactory(Favorites, {
    global: {
      plugins: [[store, key]],
    },
  });
  const setPhotosToStore = (): void => {
    store.commit(storeTypes.EMutations.ADD_TO_FAVORITES, [EXAMPLE_PHOTO]);
  };
  let dispatchSpy: jest.SpyInstance<ReturnType<Store['dispatch']>, jest.ArgsType<Store['dispatch']>>;
  let commitSpy: jest.SpyInstance<ReturnType<Store['commit']>, jest.ArgsType<Store['commit']>>;

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });

    dispatchSpy = jest.spyOn(store, 'dispatch').mockImplementation(() => new Promise((res) => {
      res(Promise.resolve([]));
    }));
    commitSpy = jest.spyOn(store, 'commit');
  });

  it('Отображение картинки, когда список пуст', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Подгрузка избранных из store', async () => {
    setPhotosToStore();

    const wrapper = getWrapper();

    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
    expect(dispatchSpy).not.toHaveBeenCalled();
  });

  it('Подгрузка избранных из локального хранилища', async () => {
    const wrapper = getWrapper();

    await flushPromises();

    expect(wrapper.find<HTMLImageElement>('img').exists()).toBe(true);
    expect(dispatchSpy).toHaveBeenCalledWith(storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE, undefined);
  });

  it('Назначается класс, когда список не пустой', async () => {
    setPhotosToStore();

    const wrapper = getWrapper();

    await flushPromises();

    expect(wrapper.element.classList).toContain('inner-container__align-start');
  });

  it('Проверка обработки события removeFromFavorites', () => {
    setPhotosToStore();

    const wrapper = getWrapper();
    const photo = wrapper.findComponent(Photo);

    photo.vm.$emit('removeFromFavorites', EXAMPLE_PHOTO);

    expect(commitSpy).toHaveBeenCalledWith(storeTypes.EMutations.REMOVE_FROM_FAVORITES, EXAMPLE_PHOTO);
  });

  it('Сохранение избранных в локальное хранилище', () => {
    const wrapper = getWrapper();

    wrapper.unmount();

    expect(dispatchSpy).toHaveBeenCalledWith(storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE, undefined);
  });
});
