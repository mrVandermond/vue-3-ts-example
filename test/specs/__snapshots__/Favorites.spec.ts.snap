// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component Отображение картинки, когда список пуст 1`] = `
<div class="inner-container">
  <div class="empty-block">
    <img src="assets/empty.png"
         class="empty-image"
         alt="empty-image"
    >
    <p class="title-text title-text__bold">
      Список избранного пуст
    </p>
    <p class="title-text">
      Добавляйте изображения, нажимая на звездочки
    </p>
  </div>
</div>
`;

exports[`Favorites component Подгрузка избранных из store 1`] = `
<div class="inner-container inner-container__align-start">
  <div class="list-photo">
    <photo-stub photo="[object Object]">
    </photo-stub>
  </div>
</div>
`;
